### PR TITLE
Adding Akamai Netstorage as a new backend.

### PR DESCRIPTION
**Pull request by aka-nil**
#### What is the purpose of this change?

This change is to add official support for Akamai Netstorage to include as a new backend.

#### Was the change discussed in an issue or in the forum before?

Adding Akamai Netstorage as a new backend [#5343](https://github.com/rclone/rclone/issues/5343)

#### Checklist

- [X] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [X] I have added tests for all changes in this PR if appropriate.
- [X] I have added documentation for the changes if appropriate.
- [X] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [X] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by ncw** [comment]
@aka-nil - we're going to need an account to integration test this with (we run this daily). Are free ones available?

Can you comment on how far you've got down this list

https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#writing-a-new-backend

In particular do the integration tests pass properly?

**Comment by aka-nil** [comment]
> @aka-nil - we're going to need an account to integration test this with (we run this daily). Are free ones available?

Credentials were sent yesterday to nick@craig-wood.com email address.

> Can you comment on how far you've got down this list
> https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#writing-a-new-backend
> In particular do the integration tests pass properly?

`test_all -backends netstorage` passed on **v1.56-stable** branch.
**TestHashSums** test fails on the master as I mentioned in the https://github.com/rclone/rclone/issues/5343

**Comment by aka-nil** [comment]
@ncw Hi Nick, how is the pull-request review going?
In the meantime we are preparing some minor bugfixing and documentation update.  I wonder what would be the procedure for this upcoming update? Do you first need to approve this pull request or we can issue a new pull request with the latest code we have?

**Comment by ncw** [comment]
@aka-nil 

Sorry, this has been on hold while we got the 1.57 release out the door.

Now it is out the door we can get on with it!

If you want to add more commits to the PR (provided the bugfixes are about this backend!) then that would be my preferred way of working.

We'll squash down to one commit when done and merge the complete working backend.

There are usually a few things which come up while merging a backend.

Can you rebase on master and fix the go1.14/go1.15 build failure please? Replace `io.NopCloser` with `ioutil.NopCloser` should do it - thanks.

File: backend/netstorage/netstorage.go
Comment (ncw) [inline backend/netstorage/netstorage.go:50]: The Help becomes command line help also.

So can you make the first sentence short and informative - this will be used for the --netstorage-protocol help. The full text will be used in the manual, so something like

```go
`Select between HTTP or HTTPS protocol.

Most users should choose HTTPS, which is the default. HTTP
is provided primarily for debugging purposes.`,

The help text should be hard wrapped at approx 80 chars also.

Same applies to all the help texts - I won't repeat this comment!

File: backend/netstorage/netstorage.go
Comment (ncw) [inline backend/netstorage/netstorage.go:58]: Do you want to set a `Default: "https"` here and remove the required? That might make more sense.

This should probably have `Advanced: true` also as it isn't needed for normal operations.

File: backend/netstorage/netstorage.go
Comment (ncw) [inline backend/netstorage/netstorage.go:69]: You want to set IsPassword if the key is the kind of thing someone could "eyedrop" lets say "mypassword" - if it is a long hex string then we don't normally encode these in rclone as they can't be memorised easily.

File: backend/netstorage/netstorage.go
Comment (ncw) [inline backend/netstorage/netstorage.go:79]: The Short help should be a single line. The Long help needs hard wrapping.

File: backend/netstorage/netstorage.go
Comment (ncw) [inline backend/netstorage/netstorage.go:212]: Normally we just put URLs in the config without separating the protocol. That would probably look neater here just to have `endpoint = https://......`

File: backend/netstorage/netstorage.go
Comment (ncw) [inline backend/netstorage/netstorage.go:213]: Debugs like this probably need to be removed before the final merge!

File: backend/netstorage/netstorage.go
Comment (ncw) [inline backend/netstorage/netstorage.go:314]: Debugf to go... You can comment it out if you want!

File: backend/netstorage/netstorage.go
Comment (ncw) [inline backend/netstorage/netstorage.go:464]: Please don't ignore errors!

File: backend/netstorage/netstorage.go
Comment (ncw) [inline backend/netstorage/netstorage.go:533]: This looks like the storage doesn't natively support ListR? If it doesn't then don't implement this method - the higher layers will do exactly this!

File: backend/netstorage/sha256reader.go
Comment (ncw) [inline backend/netstorage/sha256reader.go:11]: You don't need this abstraction - you can use a `ioutil.TeeReader` I think?

File: backend/netstorage/netstorage.go
Comment (ncw) [inline backend/netstorage/netstorage.go:878]: I recommend using the abstractions in lib/http here - it looks like this could be replaced with a call to `rest.CallXML`

Take a look at how the webdav backend works, and take a look at the `shouldRetry` function.

**Review by ncw — COMMENTED** [review]
The backend works - well done! That is quite an achievement - it isn't easy writing an rclone backend :-)

I have given this a very quick review so you've got something to be going on with.

I'd really like it to use `lib/rest` if possible rather than introducing what is essentially duplicated code. That will make the maintainers life much easier.

Are there any more of the optional methods you can implement? Without `Move` or `Copy` some important features of rclone won't work (like `--backup-dir`) so if possible that would be great.

Anyway - well done - looking forward to merging this in time for 1.58 :-)

File: backend/netstorage/netstorage.go
Comment (ncw) [reply backend/netstorage/netstorage.go:464]: Acually if it is an int, shouldn't you change the definition here?

```
	Mtime      int64   `xml:"mtime,attr"`
```

Same with size, filecount etc

```

File: backend/netstorage/netstorage.go
Comment (aka-nil) [reply backend/netstorage/netstorage.go:213]: We still prefer to have a verbose output for "-v" and "-vv" levels when we troubleshooting issue happen when the client is using rclone with Netstorage. I will go ahead and rework them, so they would make more sense.

File: backend/netstorage/netstorage.go
Comment (aka-nil) [reply backend/netstorage/netstorage.go:533]: That was a temporary solution, we have full native support for ListR coming in the next poll-request.

**Comment by aka-nil** [comment]
> Are there any more of the optional methods you can implement? Without `Move` or `Copy` some important features of rclone won't work (like `--backup-dir`) so if possible that would be great.

Netstorage public API does not currently support the server-side copy/move operations.

**Comment by aka-nil** [comment]
@ncw Hey Nick, please find the Netstorage backend pull-request update after your code review.

The major change so far was to utilize the _lib/rest_ library API and specifically _rest.CallXML_ with associated _shouldRetry_ function. That's why time I went a bit beyond the backend code and modified _lib/rest/rest.go_ to pass the request trailer required to complete the Akamai Netstorage HTTP API.

As as side note, I'm adding the modified _go.mod_ file, which brings in the _errors_ library for _errors.Wrap_ calls. Let me know if you prefer to use different approach in rclone.

**Comment by ncw** [comment]
> @ncw Hey Nick, please find the Netstorage backend pull-request update after your code review.

Great :-)

> The major change so far was to utilize the _lib/rest_ library API and specifically _rest.CallXML_ with associated _shouldRetry_ function. That's why time I went a bit beyond the backend code and modified _lib/rest/rest.go_ to pass the request trailer required to complete the Akamai Netstorage HTTP API.

Can we have the change to lib/rest in a separate commit please?

> As as side note, I'm adding the modified _go.mod_ file, which brings in the _errors_ library for _errors.Wrap_ calls. Let me know if you prefer to use different approach in rclone.

We stopped using the errors library recently in rclone and switched to the standard library based error wrapping.

So

    errors.Wrap(err, "this is a problem")

becomes

    fmt.Errorf("this is a problem: %w", err)

the special `%w` verb wraps the error so it can be unwrapped.

If you could do those two changes and rebase on master and force push, I'll give it another review - thank you :-)

I'll run the CI now so you'll see if you've got any CI errors to fixup also.

**Comment by aka-nil** [comment]
> > The major change so far was to utilize the _lib/rest_ library API and specifically _rest.CallXML_ with associated _shouldRetry_ function. That's why time I went a bit beyond the backend code and modified _lib/rest/rest.go_ to pass the request trailer required to complete the Akamai Netstorage HTTP API.
> 
> Can we have the change to lib/rest in a separate commit please?

Unfortunately, the change to use lib/rest is now baked in with other changes (like using the integers for XML serialization instead of strings).

> We stopped using the errors library recently in rclone and switched to the standard library based error wrapping.
> If you could do those two changes and rebase on master and force push, I'll give it another review - thank you :-)

I removed the dependency on _github.com/pkg/errors_ and replace with _std_ library version. Rebased on master.

> I'll run the CI now so you'll see if you've got any CI errors to fixup also.

The CI tests were all good.

**Comment by aka-nil** [comment]
> What I'd like to see is you to rebase the commit off master, then re-arrage it into two commits, one for the `lib/rest` code and one for the backend code. You'd then force push this branch.
> 
> So there would just be two commits when I look at the commits tab above.
> 
> I can do this for you if you want (I've done this sort of thing lots of times, but it is very hard if you've never done it before.)

I need your guidance on this github workflow. My understanding is that all commits are merged when doing request. It would be great if you could do it for me.

**Comment by ncw** [comment]
> > What I'd like to see is you to rebase the commit off master, then re-arrage it into two commits, one for the `lib/rest` code and one for the backend code. You'd then force push this branch.
> > So there would just be two commits when I look at the commits tab above.
> > I can do this for you if you want (I've done this sort of thing lots of times, but it is very hard if you've never done it before.)
> 
> I need your guidance on this github workflow. My understanding is that all commits are merged when doing request.

That depends on the maintainer - you have the option of squashing the commits or not.

For this PR which a) adds a backend and b) adds a feature to lib/rest I'd like to merge the unsquashed commits so the two features are seperate.

> It would be great if you could do it for me.

I've done that - I've split your commit into two one for lib/rest and one for the backend.

I'll just let the CI run though then I'll merge it.

I have yet to set up the integration testing so I'd better do that next!

**Comment by aka-nil** [comment]
> I've done that - I've split your commit into two one for lib/rest and one for the backend.

Thank you for taking care of it.

> I'll just let the CI run though then I'll merge it.

One build failed - mac_amd64. Previous CI run on my branch was successfull.

> I have yet to set up the integration testing so I'd better do that next!

The following test passed on my branch, you need to use the provided test credentials to run.
```test_all -backends netstorage```

**Review by ncw — APPROVED** [review]
Thank you! I'll merge this now and set the CI up.

You can check the results at https://pub.rclone.org/integration-tests/current/ tomorrow.

**Comment by ncw** [comment]
I set the CI up  using the credentials you sent.

Why is there a directory called `.`?

```
$ rclone lsf TestnStorage:
./
test/
```

It was giving me all sorts of weird results, but it seems to have warmed up now somehow! I tried creating a directory before and it didn't appear but now all seems ok.

**Pull request merged**
